src/mlpack/core/optimizers/adam/adam_impl.hpp

```diff
@@ -113,19 +119,44 @@ double Adam<DecomposableFunctionType>::Optimize(arma::mat& iterate)
     m *= beta1;
     m += (1 - beta1) * gradient;
 
-    v *= beta2;
-    v += (1 - beta2) * (gradient % gradient);
+    if (adaMax)
+    {
+      u *= beta2;
+      arma::mat absGradient = arma::abs(gradient);
+      // Update the exponentially weighted infinity norm.
+      for (size_t row = 0; row != iterate.n_rows; ++row)
```

Comment:
I tested for matrices of size 10000 x 10000 and `arma::max` consistently performs better by more than 35%. Here's my [test program](https://gist.github.com/ivmarkp/d549f9af02c769d5496987bf9aef12c2) and I should mention that it was a very enlightening experience :)

Pushing the changes in a couple minutes.